src/backend/distributed/planner/multi_master_planner.c

```diff
@@ -397,15 +398,31 @@ BuildDistinctPlan(Query *masterQuery, Plan *subPlan)
 	bool distinctClausesHashable = true;
 	List *distinctClauseList = masterQuery->distinctClause;
 	List *targetList = copyObject(masterQuery->targetList);
-	List *columnList = pull_var_clause_default((Node *) targetList);
+	List *columnList = NIL;
 	ListCell *columnCell = NULL;
 	bool hasDistinctAggregate = false;
 
+
 	if (IsA(subPlan, Agg))
 	{
-		return subPlan;
+		/*
+		 * We don't need to add distinct plan if all of the columns used in group by
+		 * clause also used in distinct clause and we have an aggregation plan node.
+		 */
+		if (IsGroupBySubsetOfDistinct(masterQuery))
+		{
+			return subPlan;
+		}
+
+		/*
+		 * If the previous plan is an aggregation plan, don't use aggregation functions
+		 * in the target list of the distinct plan.
+		 */
+		targetList = MasterTargetList(targetList);
```

Comment:
There are two things about this call:

1. I think we shouldn't call `MasterTargetList()`. It is very confusing to see that function called here.
2. I can see that we're doing this to handle aggregates in the target list. However, now we're iterating over the same target list almost twice (see `pull_var_clause` below). Instead, we should do that once for all target entries. What do you think about doing that like this: https://github.com/citusdata/citus/compare/another_proposal?expand=1